geode-core/src/main/java/org/apache/geode/internal/JarDeployer.java

```diff
@@ -50,17 +52,21 @@
 public class JarDeployer implements Serializable {
   private static final long serialVersionUID = 1L;
   private static final Logger logger = LogService.getLogger();
-  public static final String JAR_PREFIX_FOR_REGEX = "";
+  // The pound version scheme predates the sequenced version scheme
+  private static final Pattern POUND_VERSION_SCHEME =
+      Pattern.compile("^vf\\.gf#(?<artifact>.*)\\.jar#(?<version>\\d+)$");
+  // The sequenced version scheme predates the semantic version scheme. We use this scheme when
+  // the file name has no semantic version
+  private static final Pattern SEQUENCED_VERSION_SCHEME =
+      Pattern.compile("(?<artifact>..*)\\.v(?<version>\\d++).jar$");
+  // We use this scheme when we detect that the file has semantic version information
+  private static final Pattern SEMANTIC_VERSION_SCHEME =
```

Comment:
Please consider using the official semver regex (the "one with numbered capture groups") from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string

Reply:
this is nice, but we don't really care about major.minor.patch.tags groups, all we care is the part before the version string.

Reply:
The version string IS the major.minor.patch.tags; we aren't guaranteed to get "the part before" correctly if we aren't matching the version part correctly...
For example, `javax.inject-2.4.0-b34.jar` or `spring-oxm-4.3.18.RELEASE.jar` or `hive-cli-1.2.1+spark2.jar` all have valid semantic versions.  Is it reasonable that Geode should be able to recognize the artifact ID correctly in these cases?  Why "re-invent" the semver regex when there is already an official published regex?

However, if the intent is explicitly to do something that differs from the semver standard, an explicit comment describing the versioning scheme _we_ recognize would be helpful otherwise the constant name `SEMANTIC_VERSION_SCHEME` might give the wrong impression.

Reply:
Since semver always has three numeric fields and we want to support fewer than that I agree that it is confusing to name this SEMANTIC_VERSION_SCHEME. What would be a better name? EXPLICIT_VERSION_SCHEME? USER_VERSION_SCHEME?